### PR TITLE
Introduce a --ignore option to allow "sync" command to continue syncing even after a particular image sync fails

### PR DESCRIPTION
**Pull request by jaikiran**
Fixes https://github.com/containers/skopeo/issues/1466

As noted in the linked issue, in the current form, the `sync` command aborts as soon as any image tag has issue during copy. This ends up being a blocker if any one of the image tag out of hundreds (for a particular image) has some issue.

As suggested in the linked issue, the commit in this PR introduces a `--ignore` option to the `sync` command which allows the `sync` to continue with the rest of the images/tags. When an error occurs during the image copy, that error is logged and when the sync finally finishes, such errors contribute to `sync` being marked as failed.

I have tested this change against our setup where we were running into this issue. With this change, when `--ignore` is used, the error is now logged and the next image tag is processed and finally when the process exits you see the failure message and `sync` returns with a non-zero exit code. This is the snippet from such a run:

(there are a total to 352 tags identified for this image and the 154th one is the one having the issue)
```
....
INFO[0017] Copying image ref 154/352                     from="docker://dockerserver:5000/helloworld:release-1.0.0-SNAPSHOT-2019_Jun_18_08_58_UTC" to="docker://192.168.1.2:5000/helloworld:release-1.0.0-SNAPSHOT-2019_Jun_18_08_58_UTC"
Getting image source signatures
Copying blob fd48b0e2a9b9 skipped: already exists  
Copying blob 1fd3936b9903 skipped: already exists  
Copying blob a87126e1a510 skipped: already exists  
Copying blob 17dea04ab044 skipped: already exists  
Copying blob 44403e1d5128 [==================================>---] 66.9MiB / 72.4MiB
ERRO[0018] Error copying ref "docker://dockerserver:5000/helloworld:release-1.0.0-SNAPSHOT-2019_Jun_18_08_58_UTC"  error="writing blob: Patch http://192.168.1.2:5000/v2/helloworld/blobs/uploads/6fdba9fa-9367-4218-bccd-d33f38d7a833?_state=Pdq1xGqTyt00iB4TRvx8xypLgbyEj03Qnyi7PNSUfg57Ik5hbWUiOiJpZGVudGl0eSIsIlVVSUQiOiI2ZmRiYTlmYS05MzY3LTQyMTgtYmNjZC1kMzNmMzhkN2E4MzMiLCJPZmZzZXQiOjAsIlN0YXJ0ZWRBdCI6IjIwMjEtMTAtMDJUMDQ6MDc6MjAuMzE2NDYwNDhaIn0%3D: readfrom tcp 192.168.1.2:56274->192.168.1.2:5000: happened during read: Digest did not match, expected sha256:44403e1d5128211c7dcf598e85b608190560dc0ffac8d9df7f9ea1b270b60c12, got sha256:6b375acbd82c7ee62a88c06a4be0e1a7e4d0ba03450a792d8a932d7b3aa52fab"
```
That error gets reported and the process now moves on to next:
```
...
INFO[0018] Copying image ref 155/352                     from="docker://dockerserver:5000/helloworld:release_1.0.0s_Jun18_13_40_UTC" to="docker://192.168.1.2:5000/helloworld:release_1.0.0s_Jun18_13_40_UTC"
...
```
After the rest are synced, the final log message looks like:
```
INFO[0532] Synced 351 images from 1 sources             
FATA[0532] Sync failed due to previous reported error(s) for one or more images 
```
The exit code from that run is verified to be non-zero:
```
 echo $?
1
```

I thought of adding a test case, but it's not straightforward to trigger the error in first place (I can't yet think of a way). So I haven't added one.

P.S: This is my first time with Go. Is there any checkstyle/formatting rules that I can configure my IDE with? Everytime I did a certain change, it ended up formatting a huge part of the file where it removed certain whitespaces in various places. It's currently uses the IDE default formatter (I use GoLand from IntelliJ).

### CODE REVIEW

File: cmd/skopeo/sync.go [hunk 1]
```diff
@@ -104,6 +105,7 @@ See skopeo-sync(1) for details.
 	flags.StringVarP(&opts.destination, "dest", "d", "", "DESTINATION transport type")
 	flags.BoolVar(&opts.scoped, "scoped", false, "Images at DESTINATION are prefix using the full source image path as scope")
 	flags.BoolVarP(&opts.all, "all", "a", false, "Copy all images if SOURCE-IMAGE is a list")
+	flags.BoolVarP(&opts.continueOnError, "ignore", "", false, "Abort the sync if any image ref sync failed")
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:108]: (Absolutely non-blocking: _After_ the option name settles, consider whether  the `continueOnError` field should be named similarly to the option. OTOH `continueOnError` is a pretty great internal name, it might be better than a short-hand external option name.)

File: docs/skopeo-sync.1.md [hunk 2]
```diff
@@ -88,6 +88,9 @@ Print usage statement.
 
 **--retry-times**  the number of times to retry, retry wait time will be exponentially increased based on the number of failed attempts.
 
+**--ignore**
```
Comment (mtrmac) [inline docs/skopeo-sync.1.md:91]: Please update `completions/bash/skopeo` as well.

File: cmd/skopeo/sync.go [hunk 3]
```diff
@@ -578,7 +580,7 @@ func (opts *syncOptions) run(args []string, stdout io.Writer) error {
 		OptimizeDestinationImageAlreadyExists: true,
 		ForceManifestMIMEType:                 manifestType,
 	}
-
+	errorsPresent := false
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:583]: Non-blocking: the `imagesNumber` and `errorsPresent` values have the same lifetime and conceptual purpose, it would be nice for them to be initialized/updated/used together.

I.e., in this case, move the `imagesNumber := 0` initialization above just next to this newly added line.

**Review by mtrmac — COMMENTED** [review]
Thanks for the PR!

(See https://github.com/containers/skopeo/issues/1466#issuecomment-932783234 , this might not be the best solution for that particular situation; let’s discuss that in there. This PR is a useful feature anyway, e.g. to make as much progress as possible on unreliable network links.)

The implementation looks good; as usual, naming is the hard part :)

> P.S: This is my first time with Go. Is there any checkstyle/formatting rules that I can configure my IDE with? Everytime I did a certain change, it ended up formatting a huge part of the file where it removed certain whitespaces in various places. It's currently uses the IDE default formatter (I use GoLand from IntelliJ).

We use the language’s `gofmt` with the `-s` option, IIRC currently using the Go 1.17 version. I don’t know how to set that up in GoLand, maybe someone else can help. Anyway, the PR tests do a formatting check and this PR looks good as is already.

File: docs/skopeo-sync.1.md (hunk 2 above)
Comment (jaikiran) [reply docs/skopeo-sync.1.md:91]: Done. Updated the PR with this change.

File: cmd/skopeo/sync.go (hunk 3 above)
Comment (jaikiran) [reply cmd/skopeo/sync.go:582]: Done in latest update to this PR.

File: cmd/skopeo/sync.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vrothberg) [reply cmd/skopeo/sync.go:108]: The description of the flag suggests the inverse.  Shouldn't it be `"Do not abort [...]"` given the default value is `false`?

[the remark this answers is not in the document]
Comment (vrothberg) [reply cmd/skopeo/sync.go:108]: The Podman CLI uses `--ignore` in a number of commands but the "what" is pretty clear there. For instance, `podman rm --ignore $CONTAINER` won't error if the container doesn't exist (anymore).

In that regard `--ignore` would make sense if we're trying to copy an image that doesn't exist and make that non-fatal.

`--continue-on-error` seems to cover more than that including transient network errors.

In other words, I concur with @mtrmac :^)

[the remark this answers is not in the document]
Comment (rhatdan) [reply cmd/skopeo/sync.go:108]: I would prefer --keep-going, less characters.

[the remark this answers is not in the document]
Comment (jaikiran) [reply cmd/skopeo/sync.go]: Updated the PR to use --keep-going

Comment (jaikiran) [reply cmd/skopeo/sync.go]: Now that the PR uses --keep-going, I think calling that internal variable `keepGoing` instead of `continueOnError` is slightly better. So I've updated the PR to rename this internal member.

File: cmd/skopeo/sync.go [hunk 4]
```diff
@@ -104,6 +105,7 @@ See skopeo-sync(1) for details.
 	flags.StringVarP(&opts.destination, "dest", "d", "", "DESTINATION transport type")
 	flags.BoolVar(&opts.scoped, "scoped", false, "Images at DESTINATION are prefix using the full source image path as scope")
 	flags.BoolVarP(&opts.all, "all", "a", false, "Copy all images if SOURCE-IMAGE is a list")
+	flags.BoolVarP(&opts.keepGoing, "keep-going", "", false, "Do not abort the sync if any image ref sync failed")
```
Comment (mtrmac) [inline cmd/skopeo/sync.go:108]:
```suggestion
	flags.BoolVarP(&opts.keepGoing, "keep-going", "", false, "Do not abort the sync if any image sync failed")
```
the `ImageReference` type is a Go-level concept users don’t know (and don’t need to care) about.

Maybe “image copy” instead of “image sync”, to more clearly differentiate between “the sync” and “image sync”.

(And should that be “failed” or “fails”? I’m not a native speaker.)

---

Well, the existing output does say “ref %d/%d”… I guess I’d prefer for this new option description not to introduce another instance, and to keep the progress output as is in this PR (changing the progress output text is not what this PR is about; we can do that later separately).

(Or, any other reviewers: feel free to override me and merge as is.)

**Review by mtrmac — COMMENTED** [review]
LGTM, one tiny wording nit.

(Or, others, feel free to merge as is.)

File: cmd/skopeo/sync.go (hunk 4 above)
Comment (jaikiran) [reply cmd/skopeo/sync.go]: Done. I've updated this PR to use "image copy" and also changed "failed" to "fails" (I'm not a native English speaker either, but I think fails sounds more correct).

**Review by mtrmac — APPROVED** [review]
LGTM. Thanks again!

**Comment by jaikiran** [comment]
Thanks everyone for the reviews and the inputs.

@mtrmac, do releases happen off the main branch? If so what versions get released off that branch? The reason I ask is to decide whether I should create a PR for the 1.4 release branch containing this change so that it becomes available to us whenever the next release happens.

**Comment by rhatdan** [comment]
Releases happen off of the main branch.

**Comment by mtrmac** [comment]
… and not on an extremely regular/frequent schedule, so, @jaikiran , if you need an “official” release to happen pretty soon, please speak up and we’ll tag one.

**Comment by jaikiran** [comment]
Hello @mtrmac, this particular fix and the Makefile change that was merged some days back, both, will help us in our usage of skopeo in our product. So yes, a release will be really helpful. Thank you very much.
